src/coreclr/vm/weakreferencenative.cpp

```diff
@@ -10,1031 +10,171 @@
+    // identity of the underlying COM object (assuming that object is still alive).
     SafeComHolder<IUnknown> pTargetIdentity = nullptr;
-    if (pComWeakReference != nullptr)
-    {
-        _ASSERTE(gc.target == NULL);
-
-        GCX_PREEMP();
 
-        // Using the IWeakReference*, get ahold of the target native COM object's IInspectable*.  If this resolve fails, then we
-        // assume that the underlying native COM object is no longer alive, and thus we cannot create a new RCW for it.
-        SafeComHolderPreemp<IInspectable> pTarget = nullptr;
-        if (SUCCEEDED(pComWeakReference->Resolve(IID_IInspectable, &pTarget)))
+    // Using the IWeakReference*, get ahold of the target native COM object's IInspectable*.  If this resolve fails, then we
+    // assume that the underlying native COM object is no longer alive, and thus we cannot create a new RCW for it.
+    SafeComHolderPreemp<IInspectable> pTarget = nullptr;
+    if (SUCCEEDED(pComWeakReference->Resolve(IID_IInspectable, &pTarget)))
+    {
+        if (!pTarget.IsNull())
         {
-            if (!pTarget.IsNull())
-            {
-                // Get the IUnknown identity for the underlying object
-                SafeQueryInterfacePreemp(pTarget, IID_IUnknown, &pTargetIdentity);
-            }
+            // Get the IUnknown identity for the underlying object
+            SafeQueryInterfacePreemp(pTarget, IID_IUnknown, &pTargetIdentity);
         }
     }
 
     // If we were able to get an IUnknown identity for the object, then we can find or create an associated RCW for it.
     if (!pTargetIdentity.IsNull())
     {
+        GCX_COOP();
+        OBJECTREF rcwRef = NULL;
+        GCPROTECT_BEGIN(rcwRef);
+
         if (wrapperId != ComWrappersNative::InvalidWrapperId)
         {
             // Try the global COM wrappers
             if (GlobalComWrappersForTrackerSupport::IsRegisteredInstance(wrapperId))
             {
-                (void)GlobalComWrappersForTrackerSupport::TryGetOrCreateObjectForComInstance(pTargetIdentity, &gc.rcw);
+                (void)GlobalComWrappersForTrackerSupport::TryGetOrCreateObjectForComInstance(pTargetIdentity, &rcwRef);
             }
             else if (GlobalComWrappersForMarshalling::IsRegisteredInstance(wrapperId))
             {
-                (void)GlobalComWrappersForMarshalling::TryGetOrCreateObjectForComInstance(pTargetIdentity, ObjFromComIP::NONE, &gc.rcw);
+                (void)GlobalComWrappersForMarshalling::TryGetOrCreateObjectForComInstance(pTargetIdentity, ObjFromComIP::NONE, &rcwRef);
             }
         }
 #ifdef FEATURE_COMINTEROP
         else
         {
             // If the original RCW was not created through ComWrappers, fall back to the built-in system.
-            GetObjectRefFromComIP(&gc.rcw, pTargetIdentity);
+            GetObjectRefFromComIP(&rcwRef, pTargetIdentity);
         }
 #endif // FEATURE_COMINTEROP
+        GCPROTECT_END();
+        retRcw.Set(rcwRef);
```

Comment:
The reference needs protection as long as we may be making calls that could trigger GC. In coop mode GC can happen only if we trigger it.
`ObjectHandleOnStack::Set` does not trigger GC, so we can stop protecting before calling it.  

This can be changed as a matter of style or consistency, but I think either way is ok.

Reply:
Thanks for explaining coop mode!